ci/checks/doxygen.sh

```diff
@@ -0,0 +1,29 @@
+#!/bin/bash
+# Copyright (c) 2022, NVIDIA CORPORATION.
+###############################
+# cuDF doxygen warnings check #
+###############################
+
+# Utility to return version as number for comparison
+function version { echo "$@" | awk -F. '{ printf("%d%03d%03d%03d\n", $1,$2,$3,$4); }'; }
+
+# doxygen supported version 1.8.20 to 1.9.1
+DOXYGEN_VERSION=`doxygen --version`
+if [ $(version "$DOXYGEN_VERSION") -lt $(version "1.8.20") ] ||  [ $(version $DOXYGEN_VERSION) -gt $(version "1.9.1") ]; then
+  echo -e "Unsupported doxygen version $DOXYGEN_VERSION"
+  echo -e "Expecting doxygen version from 1.8.20 to 1.9.1"
+  exit 1
```

Comment:
>What happens if doxygen is not installed?
Fails
> What happens if an incompatible version is installed?
Fails

If doxygen-check passes when doxygen is not installed or with unsupported version, I am concerned that users might be confused that local pre-commit doxygen-check passes, but pre-commit in CI failed. 
Since pre-commit swallows STDOUT and STDERR when exit code is zero, users may not know the reason for local doxygen-check pass. IMO, it's showing different behavior in CI vs local machine silently, which ideally should not happen.

Reply:
Note that it is not possible to satisfy this version requirement (1.8.20 <= version <= 1.9.1) using `apt` on Ubuntu 18.04 or 20.04. Only Ubuntu 22.04 has a compatible version (doxygen 1.9.1) and that could be upgraded to an incompatible version in the future at any time. The lack of graceful exit means that users effectively _must_ use conda to satisfy the doxygen requirement (or build doxygen from source), which we specifically sought to avoid in the past. Users should be able to make pre-commit pass with software from only a system package manager, with no conda required. Otherwise users will simply not utilize pre-commit and benefit from the majority of features that _do_ work on all systems.

Reply:
> We already lean towards letting things pass locally and fail in CI rather than putting requirements on users to install all system dependencies

Can you please provide more details on how this decision was made? I personally find it's impractical that sometimes my local pre-commit checks passed but then the CI failed. e.g. some of the cmake format failures, if not found by the local pre-commit hook, were obvious but others took me a while to "guess" where went wrong. If users choose to contribute to the code, I think it's fair to require them to install all the proper dependencies.

> Python users in particular shouldn’t be required to install doxygen if they only make a small C++ tweak.

If they commit the "tweak", enforcing doxygen check is a reasonable requirement.

Reply:
> If they commit the "tweak", enforcing doxygen check is a reasonable requirement.

My point is that the user may not touch something that would require documentation updates, but doxygen 1.8.20 would still be required to run other checks like clang-format on commit. (Unless we rely on doxygen being last in the pre-commit configuration and tell users to skip pre-commit when running for a second time. That’s annoying.)

Reply:
(FWIW, this entire discussion becomes moot if someone packages doxygen in the same way that clang-format was packaged as a wheel. Then it could be managed by pre-commit rather than used as a system package. This is why we didn’t have clang-format in pre-commit until recently — it couldn’t be managed automatically for users.)

Reply:
Right, I totally get your point here.

There are actually pros and cons in either way (exit gracefully or not). By all means, if a PR fails doxygen check, it's still the contributor's responsibility to fix the issue so the person needs to install the proper version of doxygen anyway. 

How about exiting cleanly and adding a detailed message when doxygen check fails? e.g. providing instrucions about the proper version of doxygen, how to reproduce the doxygen error locally, etc. In this way, users won't be surprised by the unexpected CI failure and know the right way to fix it.

Reply:
> (FWIW, this entire discussion becomes moot if someone packages doxygen in the same way that clang-format was packaged as a wheel. Then it could be managed by pre-commit rather than used as a system package. This is why we didn’t have clang-format in pre-commit until recently — it couldn’t be managed automatically for users until the wheel was released.)

I actually check the clang-format wheel repo to see how much effort is needed to maintain such a project for doxygen. It requires some work to set it up, but once it's working, the maintenance seems effortless. So if it's really useful to us, we can build the "wheel" by ourselves.

Reply:
> it's still the contributor's responsibility to fix the issue so the person needs to install the proper version of doxygen anyway.

I would argue that while it is the contributor's responsibility, that doesn't imply that the contributor needs to install doxygen locally. I frequently fix problems by reading the output of CI rather than reproducing the problem locally. Failing at the "local commit" stage just means a user has more dependency installation hoops to jump through before they can commit and push to see the failing output in CI. The CI logs would direct them on what to fix locally.

Reply:
@karthikeyann @PointKernel I accept that there are pros and cons to any decision here. I approve this change but I expect pushback from developers (including myself in some cases) who do not use a conda environment and will see local failures due to introducing the first hard dependency that cannot be managed by pre-commit itself. 😕

Reply:
@bdice is correct, running the cmakelang hooks require that libcudf be configured because configuration downloads rapids-cmake into the build directory and rapids-cmake contains a format file that we need. That is the primary reason for that hook being designed to allow "soft" failures, and it's necessary because CMake allows users to place the build directory wherever they want. In theory we could _require_ users to set environment variables to configure the search, but we decided that was overkill given that the number of users who make CMake modifications beyond just adding a single new file to a CMakeLists.txt is extremely limited. With the current approach only users who actually make substantial CMake modifications need to care about this.

I agree with @PointKernel that it is reasonable to enforce documentation style on any C++ contribution, no matter how small, but I'm with Bradley that this doesn't imply a need for _local_ failures, as long as CI fails and the user is made aware of why. However, I think the real problem that we run into here is this one:

> I personally find it's impractical that sometimes my local pre-commit checks passed but then the CI failed. e.g. some of the cmake format failures, if not found by the local pre-commit hook, were obvious but others took me a while to "guess" where went wrong.

Ideally we want to provide users sufficient diagnostic information as to why pre-commit passed locally but then failed on CI. I think that the appropriate solution here is to change any hooks that we want to have this "silent pass" behavior to always be verbose (we can set `verbose: true` in the `.pre-commit-config.yaml` file for those hooks). It'll add a little bit of noise to the pre-commit run even when it passes (IIRC it'll print the hook id and how long the hook took to run), but I think that's an acceptable tradeoff for providing useful diagnostics. Currently when run verbosely the cmake-format and cmake-lint hooks will print which format file they are using, or if none was found. Always presenting that information to the user would help alleviate concerns over different behaviors locally and in CI. Similarly, including diagnostic information about doxygen versions in the doxygen hook (the script already has them, we just need to stop pre-commit from swallowing them) would indicate to users why they run into problems. This could be especially valuable for users that run manually (not through pre-commit) but are running a different local version than our required version. Running the pre-commit hook would immediately tell them why their local linter was producing output that the CI linter considers invalid.

Reply:
Brilliant idea for `verbose: true` @vyasr! That's a great solution and I like the implementation @karthikeyann. Nice teamwork!

Reply:
Done 👍 
Thank you @bdice @PointKernel @vyasr ! I really appreciate this great discussion.

Reply:
I'm not sure about how this decision played out.

To make sure I understand... 
- I want to locally reproduce a Doxygen warning failure in CI 
- I don't have the same Doxygen version installed 
- Therefore, I need to go download and install the same Doxygen version as used by CI. 

I think any solution that requires a contributor to have to go out and install some new package just to merge my PR to is a barrier to entry.

Reply:
What if we just built the right version of Doxygen from source as a dependency of the `make docs` cmake target?

Reply:
It was my understanding that the CI failure would contain enough information about the problem and solution that a local reproduction is not necessary. Is that true?

Reply:
Until recently, it was required to have a specific version of clang-format to locally reproduce a CI failure. This was simplified by using a pre-commit hook that can manage its own installation of clang-format. The doxygen requirement is no more stringent than what we previously required for clang-format.

Reply:
> It was my understanding that the CI failure would contain enough information about the problem and solution that a local reproduction is not necessary. Is that true?

I think it's impractical to expect this to be true. Imagine CI generated 10 warnings and I wanted to go through and iteratively address them. Do I need to submit to GitHub and wait for CI all 10 times to ensure each fix is working as expected?

> The doxygen requirement is no more stringent than what we previously required for clang-format.

Right, and I wasn't a fan then either. 

If we're going to require a specific version of a thing, we should automate getting it. We did this with `clang-format`. I think we can do this w/ Doxygen by just building it from source (depending on how onerous building it from source is).

Reply:
I’d support exploring the build-from-source option within pre-commit. I wonder if pre-commit hooks can execute code on installation (or build on first run?). That would be equally as good as pip-installable wheels. However, there is already a model for making wheels for non-Python packages in the clang-format wheels so I might lean towards trying that way first.

Reply:
> Imagine CI generated 10 warnings and I wanted to go through and iteratively address them. Do I need to submit to GitHub and wait for CI all 10 times to ensure each fix is working as expected?

I think all the warnings show at once. I don’t think it exits early at the first failure, so a single CI run should be enough to fix all problems. Luckily, the style check runs quickly so iteration is fast between pushing local changes and watching for CI failures.

Reply:
All the errors come at once. No early exit.

Reply:
> I’d support exploring the build-from-source option within pre-commit. 

Do you mean, build-doxygen-from-source? 
It brings doxygen's compilation problems into pre-commit installation. I would support doxygen binaries through pip-installable wheels. https://sourceforge.net/projects/doxygen/files/ is actively maintained. so it's possible to implement similar to clang-format wheels.

Reply:
I'm confused here.
> If we're going to require a specific version of a thing, we should automate getting it. We did this with clang-format.

By "we did it with clang-format", do you mean in the form of our conda environment? If so, that's still the case. Someone creating a custom environment for building cudf (or using compose) will get the correct version. If someone tries to use their own environment, though, they're on their own. That is also identical with how clang-format worked before.

> I wonder if pre-commit hooks can execute code on installation (or build on first run?). That would be equally as good as pip-installable wheels.

I don't know how much control you have over the installation step, but you can execute basically arbitrary code on first run to do the installation if it doesn't already exist (in some prespecified directory). I wouldn't say that it's equally as good, though. Compiling code within a pre-commit hook immediately opens you up to any sort of build failure, which is also essentially out of your control to debug.

> However, there is already a model for making wheels for non-Python packages in the clang-format wheels so I might lean towards trying that way first.

This is the better approach if we feel strongly about doxygen needing to always be available in this fashion. If we're already going to do the work to build it in a pre-commit hook, we're most of the way there to building the wheels.

Reply:
>"We did this with clang-format"

clang-format binary is packaged as pip installable repo. [ssciwr/clang-format-wheel](https://github.com/ssciwr/clang-format-wheel)). So it avoided using creating conda environment. 
My first idea is to use conda for automate installing specific version of doxygen. but it creates additional dependency to have `conda` installed in the system, which we don't like to put these additional conditions (`conda`) on the developer. 

@jrhemstad is suggesting to create a pip installable doxygen repo which can install doxygen binary, similar to clang-format.
@jrhemstad are you suggesting to build doxygen-from-source in pre-commit installation OR in pip installable doxygen repo releases?

Reply:
I’d echo all of what @vyasr said here: https://github.com/rapidsai/cudf/pull/11076#discussion_r921514845

Building from source on the first run of the pre-commit hook is not ideal — I only proposed exploring it because it was a different approach than previous ideas. Packaging doxygen binary wheels is the way I’d like to go if we decide to move forward here.